plugins/processors/awsapplicationsignals/internal/resolver/endpointslicewatcher.go

```diff
@@ -0,0 +1,297 @@
+	}
+}
+
+// run starts the endpointSliceWatcher.
+func (w *endpointSliceWatcher) Run(stopCh chan struct{}) {
+	w.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			w.handleSliceAdd(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			w.handleSliceUpdate(newObj, oldObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			w.handleSliceDelete(obj)
+		},
+	})
+	go w.informer.Run(stopCh)
+}
+
+func (w *endpointSliceWatcher) waitForCacheSync(stopCh chan struct{}) {
+	if !cache.WaitForNamedCacheSync("endpointSliceWatcher", stopCh, w.informer.HasSynced) {
+		w.logger.Fatal("timed out waiting for endpointSliceWatcher cache to sync")
+	}
+	w.logger.Info("endpointSliceWatcher: Cache synced")
+}
+
+// extractEndpointSliceKeyValuePairs computes the relevant mappings from an EndpointSlice.
+//
+// It returns a list of kvPair:
+//   - All IP and IP:port keys (isService=false) -> "workload@ns"
+//   - The Service name key (isService=true) -> first "workload@ns" found
+//
+// This function does NOT modify ipToWorkload or serviceToWorkload. It's purely for computing
+// the pairs, so it can be reused by both add and update methods.
+func (w *endpointSliceWatcher) extractEndpointSliceKeyValuePairs(slice *discv1.EndpointSlice) []kvPair {
+	var pairs []kvPair
+
+	for _, endpoint := range slice.Endpoints {
+		podName := ""
+		ns := ""
+		if endpoint.TargetRef != nil {
+			if endpoint.TargetRef.Kind == "Pod" {
+				podName = endpoint.TargetRef.Name
+			}
+			ns = endpoint.TargetRef.Namespace
+
+			derivedWorkload := inferWorkloadName(podName)
+			if derivedWorkload == "" {
+				w.logger.Warn("failed to infer workload name from Pod name", zap.String("podName", podName))
+				continue
+			}
+			fullWl := derivedWorkload + "@" + ns
+
+			// Build IP and IP:port pairs
+			for _, addr := range endpoint.Addresses {
+				// "ip" -> "workload@namespace"
+				pairs = append(pairs, kvPair{
+					key:       addr,
+					value:     fullWl,
+					isService: false,
+				})
+
+				// "ip:port" -> "workload@namespace" for each port
+				for _, portDef := range slice.Ports {
+					if portDef.Port != nil {
+						ipPort := fmt.Sprintf("%s:%d", addr, *portDef.Port)
+						pairs = append(pairs, kvPair{
+							key:       ipPort,
+							value:     fullWl,
+							isService: false,
+						})
+					}
+				}
+			}
+		}
+
+	}
+
+	// Service reference: "kubernetes.io/service-name"
+	svcName := slice.Labels["kubernetes.io/service-name"]
+	if svcName != "" {
+		svcFull := svcName + "@" + slice.Namespace
+
+		// If there's at least one endpoint, derive the workload from it
+		if len(slice.Endpoints) > 0 {
+			if endpoint := slice.Endpoints[0]; endpoint.TargetRef != nil {
+				derived := inferWorkloadName(endpoint.TargetRef.Name)
+				if derived != "" {
+					firstWl := derived + "@" + endpoint.TargetRef.Namespace
+					pairs = append(pairs, kvPair{
+						key:       svcFull,
+						value:     firstWl,
+						isService: true,
```

Comment:
nit: should we follow the same pattern that we used on IP to workload handling?

`if derivedWorkload == "" {
				w.logger.Warn("failed to infer workload name from Pod name", zap.String("podName", podName))
				continue
			}`

Reply:
Right. I will modify it.